rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,212 @@
+  inject: (state: PersistableState, references: SavedObjectReference[]) => PersistableState,
+  // extract function receives state and should return state with references extracted and array of references
+  // default returns same state with empty reference array
+  extract: (state: PersistableState) => [PersistableState, SavedObjectReference[]]
+}
+
+class PersistableStatePlugin {
+  setup(core, plugins) {
+    return {
+      // if one of the functions is not provided default is used 
+      // so we can always assume that all of the functions are available for every persistable state 
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
+      // takes the state, references and version and returns latest state version with references injected
+      afterLoad: (id: string, state: PersistableState, references: SavedObjectReference[], version: string) => PersistableState,
+    }
+  }
+}
+```
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const expressionReadyForSaving = persistableStatePlugin.beforeSave('expression', expressionString);
+
+``` 
+
+## EnhacedDrilldownEmbeddableInput
+
+Drilldown  plugin adds to the state of embeddable:
+
+```ts
+interface EnhancedDrilldownEmbeddableInput extends EmbeddableInput {
+  enhancements: {
+    drillDowns: { events?: Events[];  }
+  }
+}
+```
+Embeddable is aware that it has an enhancements property on its state, where every key represents state of another plugin.
+
+```ts
+const inject = (state, references) => {
+  Object.getKeys(state.enhancements).forEach(key => {
+    State.enhancements[key] = persistableStatePlugin.get(key).inject(state.enhancements[key], references);
+  }
+}
+```
+
+And drilldown plugin registers its state with persistableStateService.
+
+## We have to avoid shared persistable state
+
+So we have to do two things:
+
+- Always call a migration function for nested state.
+- Maintain state boundaries along plugin boundaries. Do not allow VisualizeEmbeddableInput to migrate data on EmbeddableInput. Don't even give it access to this state at all.
+As long as state ownership is isolated, we can do something like this:
+
+```ts
+const embeddableMigrator: PersistableStateMigrationFn (state: { type: string; input: unknown }, version: string): State {
+  return {
+    ...migrateInputToLatest(state.input),
+    enhancements: {
```

Comment:
I wonder if we should give a recommendation that the keys follow a certain pattern? Like `{pluginId}+{stateType}`?

Otherwise I think there could be key clashes that may not be known about until the "enhancer" actually registers a migration.